### ISSUE TITLE
Missing plugins and "EXDEV: cross-device link not permitted" error during plugin updates

### ISSUE THREAD

**Issue by glimmung** [saltcorn/saltcorn#2410]
**Describe the bug**
An error is thrown when I attempt to check the configuration after upgrading a plugin which suggests that the plugin has gone, and when I attempt to  re-install a previously installed plugin an explicit error is thrown.

**To Reproduce**
1. Go to 'Settings -> Modules -> Upgrade installed Modules' and click to upgrade.
2. Go to  'Settings -> System -> Restart Server' and click to restart
3. Click 'Configuration Check'
4. See errors in the report indicating that upgraded plugins have issues of the types: -
 'Field htmltype in table test has no type. Uninstalled plugin?' and 'Page Calendar: Cannot find viewtemplate Calendar in view Calendar'

**Expected behavior**
No visible change to the working system.

**Additional context**
- When I attempt to uninstall and reinstall a problem plugin, I get an error like 'EXDEV: cross-device link not permitted, rename '/tmp/saltcorn/saltcorn-tmp/temp_install/@saltcorn/page-to-pdf' -> '/home/saltcorn/plugins_folder/@saltcorn/page-to-pdf'', so it looks as though this may have been the issue during upgrades - the '/tmp/saltcorn/saltcorn-tmp/temp_install/@saltcorn/' folder contains all the upgraded plugins, not just the couple I have tried to un-install and re-install.
- Digging into the server setup, '/tmp' is indeed on a separate device to '/home/saltcorn'.
- The server setup has not changed, and this instance of Saltcorn has been running on the server for over a year, during which time I have periodically updated plugins in the same way as today.

I guess I could move the '/tmp' folder to the same device as the home folder, but I am reluctant to tinker in case I cause further problems.

**Comment by christianhugoch** [comment]
Can you say whether '/tmp' and '/home/saltcorn' are on different partitions or disks? I've looked a bit for 'EXDEV: cross-device link not permitted, and this seems to be the reason.

We've replaced the live-plugin-manager with our own plugin installer. The new installer runs a npm install in the tmp directory and moves the plugin directory over to the plugins_folder.

I believe we could add a fallback for when the move (rename) fails and do a slower copy. Windows already always uses copy.

**Comment by christianhugoch** [comment]
And can you check your npm + node version

**Comment by glimmung** [comment]
Setting the TMPDIR in the Saltcorn user's profile had no effect, but adding the setting to the "saltcorn.service" file is now working for me - I can reinstall the problem plugins now, so that seems to be a usable workaround for now.

**Comment by christianhugoch** [comment]
Okay, I think we have to add the fallback anyway, but with a workaround, it's not that urgent.

Perhaps we could add a configuration, where to run npm install, too

**Comment by christianhugoch** [comment]
thanks for spotting

**Comment by glimmung** [comment]
I agree that you need to add the fallback - this would be a show stopper for some folks.

Regarding the workaround, for anyone that needs to do that I should say that I think I misinterpreted what went on with that: -
- setting a per user temp folder for the Saltcorn user didn't appear to work.
- adding that environment variable to the service script did.
- however, tinkering with the script caused the service to reinitialise, and I don't know if that picked up the user-level change, such that the service-level change was redundant.

Thanks for the quick response!

**Comment by SaschaMn** [comment]
I have the same issue after the update but I don‘t know how to fix this, I can‘t follow your steps as linux noob.
So I restored a backup. Could you please help me what exactly to do for further updates?

**Comment by christianhugoch** [comment]
Can you try it again with [v0.9.5-beta.15](https://github.com/saltcorn/saltcorn/commit/fc46a7a5e60b8d87e8fb09d816a236c5962f0994)
